### PR TITLE
Upated for process parallelized ANTs.

### PR DESCRIPTION
**Pull request by Chris210634**
@danmcp,

I updated this plug-in to support multiple instances. It should work for any number of instances (but I haven't tested above four). 
For the plug-in to work, each instance needs to be launched with environmental variable NUMBER_OF_WORKERS=#. The memory limit also needs to be raised to 3GiB, but I'm working to fix this. 

Right now, the plug-in is working from my fork of ANTs and ITK, do we want to make a fork in FNNDSC?

### CODE REVIEW

File: antsreg/antsreg.py [hunk 1]
```diff
@@ -49,6 +63,149 @@ def define_parameters(self):
         self.add_argument('-f', dest='fixed', type=str, optional=False,
                           help='The filename for the fixed image.')
 
+    def get_process_number(self):
+        """
+        Get the next available process number.
+        Process numbers range from 0 to NUMBER_OF_PROCESSES.
```
Comment (danmcp) [inline antsreg/antsreg.py]: Shouldn't this be to NUMBER_OF_PROCESSES-1

File: antsreg/antsreg.py [hunk 2]
```diff
@@ -49,6 +63,149 @@ def define_parameters(self):
...
+        state = self.IDLE
+        while state != self.EXIT:
+            try:
+            	state = self.get_state()
+            except FileNotFoundError:
+                # Most likely because master hasn't created slave_state file yet
+                pass
+            if state == self.START:
+                self.ants_registration_command_wrapper([])
+                #Wait for IDLE signal from master
```
Comment (danmcp) [inline antsreg/antsreg.py]: Add a space after # to be consistent

File: antsreg/antsreg.py [hunk 3]
```diff
@@ -70,21 +228,53 @@ def run(self, options):
...
+        else:
+            # Slaves never return from this function.
+            self.run_parallel_ants_registration_slave();
+        # Master process must reset itkbarrier files to 0
+        for i in range(int(os.environ['NUMBER_OF_PROCESSES'])):
+            with open(self.tmp_path + '/itkbarrier' + str(i),'wb+') as barrier_file:
+                barrier_file.write(b'\0'*8) #unsigned long
+                barrier_file.close()
+
+        # variables below store abosolute file paths
```
Comment (danmcp) [inline antsreg/antsreg.py]: Capital V to be consistent

File: antsreg/antsreg.py [hunk 4]
```diff
@@ -94,26 +284,28 @@ def run(self, options):
...
 
             else:
                 filename_split = name.strip().split('.')
                 if len(filename_split) > 1 and filename_split[-1] == 'nii':   # .nii extension
                     moving_image_list.append(in_path + '/' + name)
                 elif len(filename_split) > 2 and filename_split[-2] == 'nii': # .nii.gz extension
                     moving_image_list.append(in_path + '/' + name)
-            
+
+        #Set Enviornment variable for single process exectution
```
Comment (danmcp) [inline antsreg/antsreg.py]: Space after #

File: antsreg/antsreg.py [hunk 5]
```diff
@@ -123,20 +315,17 @@ def run(self, options):
...
-                                   moving_image_name, out_path,
-                                   name_wo_ext)
-                           .split())
-            # output will be named <name_wo_ext>Warped.nii.gz
+            # Set Enviornment variable for multiple process exectution
+            os.environ['ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS'] = os.environ['NUMBER_OF_PROCESSES']
+            args = [fixed_image_name,moving_image_name,out_path,name_wo_ext]
+            self.run_parallel_ants_registration_master(args)
             
+            #Set Enviornment variable for single process exectution
```
Comment (danmcp) [inline antsreg/antsreg.py]: Space after #

File: antsreg/antsreg.py [hunk 6]
```diff
@@ -49,6 +63,149 @@ def define_parameters(self):
...
+        with self.proc_num_file_lock.acquire():
+            try:
+                with open(self.proc_num_file_path,'x') as proc_num_file:
+                    # current process is assigned 0, next process is assigned 1.
+                    proc_num_file.write('1') 
+                    proc_num_file.close()
+                    proc_num = 0
+            except FileExistsError:
+                with open(self.proc_num_file_path,'r+') as proc_num_file:
+                     # read process number and overwrite with next process number
```
Comment (danmcp) [inline antsreg/antsreg.py]: Spacing is incorrect here

**Comment by Chris210634** [comment]
Fixed comment formatting. Changed NUMBER_OF_PROCESSES to NUMBER_OF_WORKERS.

**Comment by Chris210634** [comment]
Updated with memory, cpu and number of workers min and max. ANTs still only launches one thread per worker. I'm still working on multi threading within a worker. 

@danmcp, @jbernal0019 Is this ok to merge?

**Pull request merged**